echo-pubsub/src/main/java/com/netflix/spinnaker/echo/pubsub/amazon/SQSSubscriber.java

```diff
@@ -0,0 +1,255 @@
+import com.netflix.spinnaker.echo.model.pubsub.PubsubSystem;
+import com.netflix.spinnaker.echo.pubsub.PubsubMessageHandler;
+import com.netflix.spinnaker.echo.pubsub.model.PubsubSubscriber;
+import com.netflix.spinnaker.echo.pubsub.utils.NodeIdentity;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * One subscriber for each subscription.
+ * The subscriber makes sure the SQS queue is created, subscribes to the SNS topic,
+ * polls the queue for messages, and removes them once processed.
+ */
+public class SQSSubscriber implements Runnable, PubsubSubscriber {
+
+  private static final Logger log = LoggerFactory.getLogger(SQSSubscriber.class);
+
+  private static final int AWS_MAX_NUMBER_OF_MESSAGES = 10;
+  static private final PubsubSystem pubsubSystem = PubsubSystem.AMAZON;
+
+  private ObjectMapper objectMapper;
+  private AmazonSNS amazonSNS;
+  private AmazonSQS amazonSQS;
+
+  private AmazonPubsubProperties.AmazonPubsubSubscription subscription;
+
+  private PubsubMessageHandler pubsubMessageHandler;
+  private MessageArtifactTranslator messageArtifactTranslator;
+
+  private NodeIdentity identity = new NodeIdentity();
+
+  Registry registry;
+
+  private final ARN queueARN;
+  private final ARN topicARN;
+
+  private String queueId = null;
+
+
+  public SQSSubscriber(ObjectMapper objectMapper,
+                       AmazonPubsubProperties.AmazonPubsubSubscription subscription,
+                       PubsubMessageHandler pubsubMessageHandler,
+                       AmazonSNS amazonSNS,
+                       AmazonSQS amazonSQS,
+                       Registry registry) {
+    this.objectMapper = objectMapper;
+    this.subscription = subscription;
+    this.pubsubMessageHandler = pubsubMessageHandler;
+    this.amazonSNS = amazonSNS;
+    this.amazonSQS = amazonSQS;
+    this.registry = registry;
+
+    this.messageArtifactTranslator = new MessageArtifactTranslator(subscription.readTemplatePath());
+    this.queueARN = new ARN(subscription.getQueueARN());
+    this.topicARN = new ARN(subscription.getTopicARN());
+  }
+
+  public String getWorkerName() {
+    return queueARN.arn + "/" + SQSSubscriber.class.getSimpleName();
+  }
+
+  @Override
+  public PubsubSystem pubsubSystem() {
+    return pubsubSystem;
+  }
+
+  @Override
+  public String subscriptionName() {
+    return subscription.getName();
+  }
+
+  @Override
+  public String getName() {
+    return getWorkerName();
+  }
+
+  @Override
+  public void run() {
+    log.info("Starting " + getWorkerName());
+
+    while (true) {
+      try {
+        listenForMessages();
+      } catch (Throwable e) {
+        log.error("Unexpected error running " + getWorkerName() + ", restarting", e);
+      }
+    }
+  }
+
+  private void listenForMessages() {
```

Comment:
Could break up the logical chunks into helpers to make this function more readable.